File: tdfc2/src/cpp/stdlib/new.cc
/* $Id$ */

/*
 * Copyright 2011, The TenDRA Project.
 * Copyright 1997, United Kingdom Secretary of State for Defence.
 *
 * See doc/copyright/ for the full copyright terms.
 */

#include "implement.h"

/*
 * The default new handler throws a bad_alloc exception.
 */
static void
default_new_handler()
{
	throw bad_alloc();
}

/*
 * This value gives the current new handler function.
 */
static new_handler crt_new_handler = default_new_handler;

/*
 * Set the current new handler function to fn, returning
 * the previous value.
 */
new_handler
std::set_new_handler(new_handler f) throw ()
{
	new_handler g = crt_new_handler;
	crt_new_handler = f;
	return g;
}

/*
 * Standard memory allocation function
 *
 * Allocate sz bytes of memory, throwing an exception if this is not possible.
 */
void *
__TCPPLUS_new(size_t sz) throw (bad_alloc)
{
	void *p;

	if (sz == 0) {
		sz = 1;
	}

	while (p = malloc(sz), p == NULL) {
		new_handler f = crt_new_handler;
		if (f == NULL) {
			throw bad_alloc();
		}

		f();
	}

	return p;
}

/*
 * No-exception memory allocation function
 *
 * This routine allocates sz bytes of memory, returning a null pointer if
 * this is not possible.
 */
void *
__TCPPLUS_new_nothrow(size_t sz) throw ()
{
	void *p;

	if (sz == 0) {
		sz = 1;
	}

	while (p = malloc(sz), p == NULL) {
		try {
			new_handler f = crt_new_handler;
			if (f == NULL) {
				return NULL;
			}

			f();
		} catch (const bad_alloc &) {
			// Return null if handler throws an exception
			return NULL;
		}
	}

	return p;
}

/*
 * Deallocate the memory given by p.
 */
void
__TCPPLUS_delete(void *p) throw ()
{
	if (p) {
		free(p);
	}

	return;
}

/*
 * The standard nothrow object
 *
 * This object is used in the no-exception memory allocation routines.
 */
const nothrow_t std::nothrow = { };

File: tspec/extension/cpp/new.ts
# Copyright 2002-2011, The TenDRA Project.
# Copyright 1997, United Kingdom Secretary of State for Defence.
#
# See doc/copyright/ for the full copyright terms.


$LINKAGE = "C++" ;
$NAMESPACE = "" ;

+USE "cpp", "cstdlib.ts" ;
+USE "cpp", "exception.ts" ;

+SUBSET "bad_alloc" := {
    # This is done as a subset because it is in the std namespace
    $LINKAGE = "C++" ;
    $NAMESPACE = "std" ;
%%
// EXCEPTION THROWN BY BAD NEW OPERATION
class bad_alloc : public exception {
    // Implementation dependent details
public :
    bad_alloc () throw () ;
    bad_alloc ( const bad_alloc & ) throw () ;
    bad_alloc &operator= ( const bad_alloc & ) throw () ;
    virtual ~bad_alloc () throw () ;
    virtual const char *what () const throw () ;
} ;

// SURROGATE NO-EXCEPTION CLASS
struct nothrow_t {} ;
extern const nothrow_t nothrow ;

// NEW HANDLERS
typedef void ( *new_handler ) () ;
new_handler set_new_handler ( new_handler ) throw () ;
%%
} ;

%%
// STANDARD ALLOCATION OPERATORS
void *operator new ( std::size_t ) throw ( std::bad_alloc ) ;
void *operator new [] ( std::size_t ) throw ( std::bad_alloc ) ;
void operator delete ( void * ) throw () ;
void operator delete [] ( void * ) throw () ;

// NO-EXCEPTION ALLOCATION OPERATORS
void *operator new ( std::size_t, const std::nothrow_t & ) throw () ;
void *operator new [] ( std::size_t, const std::nothrow_t & ) throw () ;
void operator delete ( void *, const std::nothrow_t & ) throw () ;
void operator delete [] ( void *, const std::nothrow_t & ) throw () ;

// PLACEMENT ALLOCATION OPERATORS
inline void *operator new ( std::size_t, void *p ) throw () { return ( p ) ; }
inline void *operator new [] ( std::size_t, void *p ) throw () { return ( p ) ; }
inline void operator delete ( void *, void * ) throw () { return ; }
inline void operator delete [] ( void *, void * ) throw () { return ; }
%%
